File: Scraper/drivers/mhg.py
import json
import requests
from bs4 import BeautifulSoup
import lzstring
import time
import re

from models.driver import Driver
from models.manga import *

lz = lzstring.LZString()


def get(url, proxy):
    try:
        res = requests.get(
            url,
            timeout=5,
            proxies={"https": proxy},
        )
    except:
        return False
    m = re.match(r"^.*\}\(\'(.*)\',(\d*),(\d*),\'([\w|\+|\/|=]*)\'.*$", res.text)
    return packed(
        m.group(1),
        int(m.group(2)),
        int(m.group(3)),
        lz.decompressFromBase64(m.group(4)).split("|"),
    )


# parse.py
def packed(functionFrame, a, c, data):
    def e(innerC):
        return ("" if innerC < a else e(int(innerC / a))) + (
            chr(innerC % a + 29) if innerC % a > 35 else tr(innerC % a, 36)
        )

    c -= 1
    d = {}
    while c + 1:
        d[e(c)] = e(c) if data[c] == "" else data[c]
        c -= 1
    pieces = re.split(r"(\b\w+\b)", functionFrame)
    js = (
        "".join([d[x] if x in d else x for x in pieces])
        .replace("\\'", "'")
        .replace('\\"', '"')
    )
    return json.loads(re.search(r"^.*\((\{.*\})\).*$", js).group(1))


# tran.py
def itr(value, num):
    d = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    return "" if value <= 0 else itr(int(value / num), num) + d[value % num]


def tr(value, num):
    tmp = itr(value, num)
    return "0" if tmp == "" else tmp


class MHG(Driver):
    identifier = "MHG"
    timeout = 10
    initId = 1
    categories = {
        "rexue": Manga.categories_list[0],
        "aiqing": Manga.categories_list[1],
        "xiaoyuan": Manga.categories_list[2],
        "baihe": Manga.categories_list[3],
        "danmei": Manga.categories_list[4],
        "maoxian": Manga.categories_list[5],
        "hougong": Manga.categories_list[6],
        "kehuan": Manga.categories_list[7],
        "zhanzheng": Manga.categories_list[8],
        "xuanyi": Manga.categories_list[9],
        "tuili": Manga.categories_list[10],
        "gaoxiao": Manga.categories_list[11],
        "mohuan": Manga.categories_list[12],
        "mofa": Manga.categories_list[13],
        "kongbu": Manga.categories_list[14],
        "shengui": Manga.categories_list[15],
        "lishi": Manga.categories_list[16],
        "jingji": Manga.categories_list[18],
        "jizhan": Manga.categories_list[20],
        "weiniang": Manga.categories_list[22],
    }

    def get(self, id: int, proxy: str):
        response = requests.get(
            f"https://tw.manhuagui.com/comic/{id}/",
            proxies={"https": proxy},
            timeout=5,
        )

        if response.status_code != 200:
            raise

        soup = BeautifulSoup(response.text, "lxml")

        thumbnail = soup.find("p", class_="hcover")
        is_ended = "finish" in thumbnail.find_all("span")[-1]["class"]
        thumbnail = "https:" + thumbnail.find("img")["src"]

        title = soup.find("div", class_="book-title").find("h1").text.strip()
        info = soup.find("ul", class_="detail-list cf").find_all("li")
        categories = [
            MHG.categories[i["href"][6:-1]]
            for i in info[1].find("span").find_all("a")
            if i["href"][6:-1] in MHG.categories.keys()
        ]
        authors = [i.text.strip() for i in info[1].find_all("span")[1].find_all("a")]
        description = soup.find("div", id="intro-cut").text.strip()
        latest = soup.find("div", class_="chapter-bar").find("a").text.strip()

        tryAdult = soup.find("input", id="__VIEWSTATE")
        if tryAdult:
            chapter_list = BeautifulSoup(
                lzstring.LZString().decompressFromBase64(tryAdult.attrs.get("value")),
                "lxml",
            ).find_all("div", class_="chapter-list")
        else:
            chapter_list = soup.find_all("div", class_="chapter-list")

        def extract_chapter(raw):
            try:
                chapters = []
                for i in reversed(raw.find_all("ul")):
                    for j in i.find_all("a"):
                        chapters.append(
                            Chapter(
                                title=j["title"].strip(),
                                id=re.search("(\\d+)\\.html", j["href"]).group(1),
                            )
                        )
                return chapters
            except:
                return []

        serial = extract_chapter(chapter_list[0])
        extra = []
        for i in chapter_list[1:]:
            extra.extend(extract_chapter(i))

        manga = Manga(
            id=str(id),
            chapters=Chapters(serial=serial, extra=extra, extra_data=str(id)),
            thumbnail=thumbnail,
            title=title,
            latest=latest,
            authors=authors,
            description=description,
            is_ended=is_ended,
            categories=categories,
            update_time=time.time(),
        )

        return manga

    def next(self, id: int):
        return id + 1

    def get_update(self, proxy: str):
        response = requests.get(
            "https://tw.manhuagui.com/update/d1.html",
            proxies={"https": proxy},
            timeout=5,
        )
        soup = BeautifulSoup(response.text, "lxml")
        update = soup.find("div", class_="latest-cont").find_all("li")

        result = []
        for i in update:
            id = re.search("\/(\d+)\/", i.find("a")["href"]).group(1)
            latest = (
                i.find("span", class_="tt")
                .text.replace("更新至", "")
                .replace("共", "")
                .strip()
            )

            result.append(PreviewManga(id=id, latest=latest))

        return result

    def get_chapter(self, id, extra_data, proxy):
        details = get(f"https://tw.manhuagui.com/comic/{extra_data}/{id}.html", proxy)
        urls = list(
            map(
                lambda x: f"https://i.hamreus.com{details['path']}{x}", details["files"]
            )
        )

        return urls

    def is_same(self, val1, val2):
        return val1.replace("第", "") == val2.replace("第", "")

File: Scraper/src/drivers/mhg.ts
import { decompressFromBase64 } from "lz-string";
import fetch from "node-fetch";
import { HTMLElement, parse } from "node-html-parser";

import { Manga as MangaModel } from "@prisma/client";

import Driver from "../models/driver";
import Manga, { Chapter, Genre } from "../models/manga";

interface PreviewManga {
  id: string;
  latest: string;
}

class MHG extends Driver<PreviewManga> {
  id = "MHG";
  timeout = 5000;
  numPerPage = 42;
  static baseUrl = "https://tw.manhuagui.com";
  static imageBaseUrl = "https://i.hamreus.com";
  static genresMapping: { [key: string]: Genre } = {
    rexue: Genre.HotBlooded,
    aiqing: Genre.Romance,
    xiaoyuan: Genre.Campus,
    baihe: Genre.Yuri,
    danmei: Genre.BL,
    maoxian: Genre.Adventure,
    hougong: Genre.Harem,
    kehuan: Genre.SciFi,
    zhanzheng: Genre.War,
    xuanyi: Genre.Suspense,
    tuili: Genre.Speculation,
    gaoxiao: Genre.Funny,
    mohuan: Genre.Fantasy,
    mofa: Genre.Magic,
    kongbu: Genre.Horror,
    shengui: Genre.Ghosts,
    lishi: Genre.Historical,
    jingji: Genre.Sports,
    jizhan: Genre.Mecha,
    weiniang: Genre.Otokonoko,
  };

  async fetch(url: string): Promise<HTMLElement> {
    const resp = await fetch(`${MHG.baseUrl}${url}`, { agent: this.agent });
    const doc = parse(await resp.text());

    if (!resp.ok) throw new Error(`Failed to fetch. Code ${resp.status}`);

    return doc;
  }

  async get(id: string): Promise<Manga> {
    const doc = await this.fetch(`/comic/${id}`);

    const thumbnailContainer = doc.querySelector("p.hcover")!;
    const isEned = thumbnailContainer
      .getElementsByTagName("span")
      .at(-1)!
      .classList.contains("finish");
    const thumbnail = `https:${thumbnailContainer
      .getElementsByTagName("img")[0]
      .getAttribute("src")!}`;

    const title = doc.querySelector("div.book-title > h1")!.textContent.trim();

    const infoContainer = doc.querySelector(
      "ul.detail-list.cf > li:nth-child(2)"
    )!;

    const genres: Genre[] = [];
    for (const genre of infoContainer.querySelectorAll("span:first-child a")) {
      const match = genre.getAttribute("href")?.match(/\/list\/(.*)\//);
      if (match && MHG.genresMapping[match[1]])
        genres.push(MHG.genresMapping[match[1]]);
    }
    const authors = infoContainer
      .querySelectorAll("span:last-child a")
      .map((e) => e.textContent.trim());
    const description = doc.getElementById("intro-cut")!.textContent.trim();
    const latest = doc.querySelector("div.chapter-bar a")!.textContent.trim();
    const updateTime = Math.floor(
      new Date(
        doc
          .querySelectorAll("div.chapter-bar > span > span")[1]
          .textContent.trim()
      ).getTime() / 1000
    );

    let chaptersList = doc.querySelectorAll("div.chapter-list");

    const tryAdult = doc.querySelector("input#__VIEWSTATE");
    if (tryAdult) {
      chaptersList = parse(
        decompressFromBase64(tryAdult.getAttribute("value")!).toString()
      ).querySelectorAll("div.chapter-list");
    }

    function extractChapters(elem: HTMLElement): Array<Chapter> {
      const chapters: Array<Chapter> = [];

      for (const ul of elem.getElementsByTagName("ul").reverse()) {
        for (const link of ul.getElementsByTagName("a")) {
          try {
            chapters.push({
              title: link.getAttribute("title")!.trim(),
              id: link
                .getAttribute("href")!
                .match(/\/comic\/.*\/(.*).html/)![1],
            });
          } catch {}
        }
      }

      return chapters.reverse();
    }

    const serial = extractChapters(chaptersList[0]);
    chaptersList.shift();
    const extra = chaptersList
      .map((v) => extractChapters(v))
      .reduce((prev, v) => [...prev, ...v], []);

    return new Manga(
      id,
      title,
      authors,
      genres,
      description,
      thumbnail,
      isEned,
      latest,
      updateTime,
      { serial, extra, extraData: id }
    );
  }

  async getTotal() {
    const doc = await this.fetch("/list");

    return Number(
      doc.querySelector("div.result-count strong:last-child")!.textContent
    );
  }

  async getAll(page: number) {
    const doc = await this.fetch(`/list/index_p${page + 1}.html`);

    return doc
      .querySelectorAll("#contList a.bcover")
      .map((v) => v.getAttribute("href")!.match(/\/comic\/(.*)\//)![1]);
  }

  async getUpdates() {
    const doc = await this.fetch("/update/d30.html");

    return doc.querySelectorAll("div.latest-list li").map((v) => ({
      id: v
        .querySelector("a.cover")!
        .getAttribute("href")!
        .match(/\/comic\/(.*)\//)![1],
      latest: v.querySelector("a.cover")!.textContent.replace(/更新至|共/, ""),
    }));
  }

  async getChapter(id: string, extraData: string) {
    const doc = await this.fetch(`/comic/${extraData}/${id}.html`);

    const raw = doc.querySelector("script:not([src])[type]")!.textContent;
    const rawCode = raw.match(/(\(.*\))/)![1];

    // Replace splic with parsed and split
    const regex = new RegExp(
      /\'([-A-Za-z0-9+\/]*={0,3})\'\['\\x73\\x70\\x6c\\x69\\x63'\]/
    );
    const match = rawCode.match(regex)!;
    const replaced = rawCode.replace(
      regex,
      `${JSON.stringify(decompressFromBase64(match[1]))}.split`
    );

    // extract the arguments
    const parsed = (0, eval)(replaced) as string;
    const json = JSON.parse(parsed.match(/\((\{.*\})\)/)![1]);

    return json.files.map((v: string) => `${MHG.imageBaseUrl}${json.path}${v}`);
  }

  isEqual(newManga: PreviewManga, oldManga: MangaModel) {
    function normalize(str: string) {
      return str.replace("/第| .*/g", "").trim();
    }

    return normalize(oldManga.latest) === normalize(newManga.latest);
  }

  new() {
    return new MHG();
  }
}

export default MHG;
